flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java

```diff
@@ -417,12 +417,55 @@ public void cancel() {
 	@Override
 	public final boolean updateTaskExecutionState(final TaskExecutionState taskExecutionState) {
 		final Optional<ExecutionVertexID> executionVertexId = getExecutionVertexId(taskExecutionState.getID());
-		if (executionVertexId.isPresent()) {
-			executionGraph.updateState(taskExecutionState);
-			updateTaskExecutionStateInternal(executionVertexId.get(), taskExecutionState);
+
+		boolean updateSuccess = executionGraph.updateState(taskExecutionState);
+
+		if (updateSuccess && executionVertexId.isPresent()) {
+			Optional<TaskExecutionState> validTaskExecutionState = getValidTaskExecutionState(
+				executionVertexId.get(),
+				taskExecutionState);
+
+			if (validTaskExecutionState.isPresent()) {
+				updateTaskExecutionStateInternal(executionVertexId.get(), validTaskExecutionState.get());
+			}
 			return true;
+		} else {
+			return false;
+		}
+	}
+
+	private Optional<TaskExecutionState> getValidTaskExecutionState(
+			final ExecutionVertexID executionVertexId,
+			final TaskExecutionState taskExecutionState) {
+
+		final ExecutionVertex executionVertex = getExecutionVertex(executionVertexId);
+		final ExecutionAttemptID currentExecutionAttemptID = executionVertex.getCurrentExecutionAttempt().getAttemptId();
+
+		// check whether this state update is outdated
+		if (!currentExecutionAttemptID.equals(taskExecutionState.getID())) {
```

Comment:
At the moment, this check cannot be true, I think.

Reply:
You are right. The resetting will happen async in `DefaultScheduler#restartTasksWithDelay`, so that the execution attempt can not be outdated here.
Will remove this check then.

Reply:
done.